### DISCUSSION TITLE
How do I properly format html with indention/whitespace?

### DISCUSSION THREAD

**Discussion by j4w8n** [rehypejs/rehype#161] [Q&A] [answered]
Thank you for all of these ecosystem utilities! I'm new to them and can't figure out what I'm doing wrong here.

Input html
```html
<form><label></label><fieldset><label></label><input><label></label><input></fieldset></form>
```
Expected output
```html
<form>
  <label></label>
  <fieldset>
    <label></label>
    <input>
    <label></label>
    <input>
  </fieldset>
</form>
```

Trying this
```js
const vfile = await rehype()
  .data('settings', { fragment: true })
  .use(rehypeFormat, { indentInitial: false })
  .process('<form><label></label><fieldset><label></label><input><label></label><input></fieldset></form>')

console.log(vfile.value)
```
Actual output
```html
<form><label></label>
<fieldset><label></label><input><label></label><input></fieldset>
</form>
```

**Comment by wooorm** [comment, marked as answer]
Whitespace has meaning / is visible in certain cases in HTML. `rehype-format` can’t add/remove whitespace in all places.

**Reply by j4w8n** [reply]
Ok thanks!

**Comment by j4w8n** [comment]
Ok, so it is possible to represent what we want in `hast`; it's just really verbose. Then, once that's transformed to html, I get the result I'm looking for.

```js
  const oghtml = 
`<form>
  <label></label>
  <fieldset>
    <label></label>
    <input>
    <label></label>
    <input>
  </fieldset>
</form>`

  /* Transform to hast. */
  const hast = fromHtml(oghtml, { fragment: true })
  console.log(JSON.stringify(hast, null, 2))

  /* And back to the desired html structure. */
  const html = toHtml(hast)
  console.log(html)
```

**Reply by wooorm** [reply]
Yes, this project works on hast. So it can generate that too. But it needs to know there’s *some* space/newline between things. If you put `{type: 'text', value: '\n'}` between things, it can indent nicely: https://github.com/rehypejs/rehype-format/blob/main/test/fixtures/aria/input.html
